std/process.d

```diff
@@ -3169,33 +3169,73 @@ static:
     }
 
 private:
-    // Returns the length of an environment variable (in number of
-    // wchars, including the null terminator), or 0 if it doesn't exist.
-    version (Windows)
-    int varLength(LPCWSTR namez) @trusted nothrow
-    {
-        return GetEnvironmentVariableW(namez, null, 0);
-    }
-
     // Retrieves the environment variable, returns false on failure.
     bool getImpl(in char[] name, out string value) @trusted
     {
         version (Windows)
         {
+            // first we ask windows how long the environment variable is,
+            // then we try to read it in to a buffer of that length. Lots
+            // of error conditions because the windows API is nasty.
+
             import std.conv : to;
             const namezTmp = name.tempCStringW();
-            immutable len = varLength(namezTmp);
-            if (len == 0) return false;
+            WCHAR[] buf;
+
+            // clear error because GetEnvironmentVariable only says it sets it
+            // if the environment variable is missing, not on other errors.
+            SetLastError(NO_ERROR);
+            // len includes terminating null
+            immutable len = GetEnvironmentVariableW(namezTmp, null, 0);
+            if (len == 0)
+            {
+                immutable err = GetLastError();
+                if (err == ERROR_ENVVAR_NOT_FOUND)
+                    return false;
+                else // some other windows error. Might actually be NO_ERROR, because
+                     // GetEnvironmentVariable doesn't specify whether it sets on all
+                     // failures
+                    throw new Exception(sysErrorString(err));
+            }
             if (len == 1)
             {
                 value = "";
                 return true;
             }
+            buf.length = len;
 
-            auto buf = new WCHAR[len];
-            GetEnvironmentVariableW(namezTmp, buf.ptr, to!DWORD(buf.length));
-            value = toUTF8(buf[0 .. $-1]);
-            return true;
+            // try twice in case variable grew in the tiny window since we queried the length
```

Comment:
I was concerned about DOS. If someone could trigger a huge torrent of environment setting then other threads/processes reading env in this function could be made to wait inside std.process which isn't controllable from the user's code (as opposed to an exception after a small number of tries, at which point they can try again if they want).

I'm not that experienced with how the windows thread/process scheduler works, but if I had to guess, being usurped from the CPU would more likely to happen during a call to GetEnvironmentVariable than elsewhere (as it could involve IO or suchlike), making such an attack possible.

If you think that's all just FUD then I'd be happy to change it to infinite, or maybe just a larger number.